app/models/medium.rb

```diff
@@ -336,15 +336,12 @@ def self.search_by(search_params, _page)
         with(:release_state, search_params[:access])
       end
     end
-    if !search_params[:all_tags] == "1" &&
-       !search_params[:tag_operator] == "or" && (search_params[:tag_ids])
-      if search_params[:tag_operator] == "or" || search_params[:all_tags] == "1"
```

Comment:
We had `search_params[:all_tags] == "1"` beforehand to force when we should use a logical OR. I think this makes sense: when a user wants to include "all tags" in their search, why would they want an AND to chain them since its very improbable that every search result will have all tags included. So, I think only OR makes sense here and the user should not even be able to select "AND" in this case. Of course, this was poorly or not at all communicated to the user before this change: even when the user selected "AND" it was still an "OR" in the code here. You can test this at `git checkout b3b7f6ac96`.

If you agree, I could go ahead and disable the OR/AND once the user clicked on "all tags". But this should still be reflected in the backend then, that is, if all tags are selected always use a logical "OR".

Reply:
Yes, I agree. The OR/AND should be removed in the fronend if all tags are selected. In the frontend I think it should be sufficient to remove lines 304 to 306 in `medium.rb`. Then only the code inside

```ruby
if search_params[:all_tags] == "0" && search_params[:tag_ids].any?
  search.build do
    if search_params[:tag_operator] == "and"
      with(:tag_ids).all_of(search_params[:tag_ids])
    else
      with(:tag_ids).any_of(search_params[:tag_ids])
    end
  end
end
```

sets any restriction on the SOLR search regarding tags. If `search_params[:all_tags] == "1"` then no such restrictions meaning that the result does not take into account which tags are involved.
If we agree on this please go ahead and make the frontend (and minor backend) change. If not, we can further discuss.

Reply:
You mean these lines?

https://github.com/MaMpf-HD/mampf/blob/469a5836c67ca01873805daf0e6fd6e870539958/app/models/medium.rb#L304-L306

What is `search_params[:tag_ids] = Tag.pluck(:id)` supposed to do in the first place?

Reply:
Unfortunately, the search form has been copied in the code in some occurences, see for example the following string:

```rb
<%= f.label :tag_ids,
                  t('basics.related_tags'),
                  class: 'form-label' %>
```

in `app/views/main/start` as well as `app/views/media/catalog` and the respective complete formula that is copied.

What I struggle with is to find out the purpose of the `_search_results.html.erb` file, especially since it renders the `tags_form` where I would have to also reflect this change:

https://github.com/MaMpf-HD/mampf/blob/469a5836c67ca01873805daf0e6fd6e870539958/app/views/media/catalog/_search_results.html.erb#L105

This form is only shown if the purpose is either "media" or "clicker" as seen here:

https://github.com/MaMpf-HD/mampf/blob/469a5836c67ca01873805daf0e6fd6e870539958/app/views/media/catalog/_search_results.html.erb#L95

Search for the string `search_results"` and the only occurrence I get is in the `import_preview.coffee` file where the purpose is set. Do you know where in the UI I can find the place where `purpose` will be set to "media" such that I can see the `tags_form.html.erb` rendered?

Reply:
> What is search_params[:tag_ids] = Tag.pluck(:id) supposed to do in the first place? Apart from that I agree: if tag_ids is empty, we won't construct any restrictions, i.e. "all tags" are included.

It replaces the the tag_ids array (which is empty in this case) by the array of all ids of tags that exists. As you already wrote, it is highly unlikely that a medium exists whose associated tags are all existing tags. So, using this `tag_ids` array in combination with  `with(tag_ids).all_of(search_params[:tag_ids])` will result in no search results and using it with `with(tag_ids).any_of(search_params[:tag_ids])` results in media that have at least one (arbitrary) tag. I do not see why wee would need any of this.

Reply:
> Search for the string search_results" and the only occurrence I get is in the import_preview.coffee file where the purpose is set. Do you know where in the UI I can find the place where purpose will be set to "media" such that I can see the tags_form.html.erb rendered?

You can see `import_preview.coffee` in action in two places: For displaying search results for importing media in a lecture in a lecture edit page, see e.g. `http://localhost:3000/lectures/1/edit` in the accordion in the development setting, or for quizzes after you import questions for a quiz: `http://localhost:3000/quizzes/16/edit` (click on 'new vertex', then import question). In these cases the purpose is `quiz`/`import` so the `search_results` partial will not render the `tags_form` partial.

Reply:
> It replaces the the tag_ids array (which is empty in this case) by the array of all ids of tags that exists.

Thanks for the clarification, I know get what the purpose of this line was.

>  I do not see why wee would need any of this.

Fair enough, so I removed the logic in 7dd388313f

> the `search_results` partial could be hugely simplified.

Ok yeah, we should really fix this code duplication in some future PR to remove this dreadful sin from the code ;) Especially since these search forms are a perfect fit for reusability across multiple places.
At least, I did not have to mirror the UI change as the AND/OR fields both have the same ID in the two files they occur in, so one JS file was sufficient to handle both cases. The only real thing I had to mirror is to add the `disabled: true` statement on the respective AND/OR tag operators to disable them initially since "all" is selected initially.

Reply:
From my side, this can be resolved.